crates/aptos/src/move_tool/mod.rs

```diff
@@ -289,31 +301,42 @@ impl CliCommand<TransactionSummary> for PublishPackage {
     }
 
     async fn execute(self) -> CliTypedResult<TransactionSummary> {
-        let build_config = BuildConfig {
-            additional_named_addresses: self.move_options.named_addresses(),
-            generate_abis: false,
-            generate_docs: true,
-            install_dir: self.move_options.output_dir.clone(),
-            ..Default::default()
-        };
-        let package = compile_move(build_config, self.move_options.get_package_dir()?.as_path())?;
-        let compiled_units: Vec<Vec<u8>> = package
-            .root_compiled_units
-            .iter()
-            .map(|unit_with_source| {
-                unit_with_source
-                    .unit
-                    .serialize(get_bytecode_version_from_env())
-            })
-            .collect();
-
-        // Send the compiled module
-        self.txn_options
-            .submit_transaction(TransactionPayload::ModuleBundle(ModuleBundle::new(
+        let PublishPackage {
+            move_options,
+            txn_options,
+            new_flow,
+            upgrade_policy,
+        } = self;
+        let package = BuiltPackage::build(move_options, true, true)?;
+        let compiled_units = package.extract_code();
+        if !new_flow {
+            if upgrade_policy.is_some() {
+                return Err(CliError::CommandArgumentError(
+                    "`--upgrade-policy` can only be used with the `--new-publish` flow".to_owned(),
+                ));
+            }
+            // Send the compiled module using a module bundle
+            txn_options
+                .submit_transaction(TransactionPayload::ModuleBundle(ModuleBundle::new(
+                    compiled_units,
+                )))
+                .await
+                .map(TransactionSummary::from)
+        } else {
+            // Send the compiled module and metadata using the code::publish_package_txn.
+            let metadata = package.extract_metadata(
+                package.name().to_owned(),
+                upgrade_policy.unwrap_or_else(UpgradePolicy::compat),
+            )?;
+            let payload = aptos_transaction_builder::aptos_stdlib::code_publish_package_txn(
+                bcs::to_bytes(&metadata).expect("PackageMetadata has BCS"),
```

Comment:
Let's not do any expects, and just do `?` to convert it to errors.  Avoiding unwraps and expects in the CLI space as it doesn't end up bubbling up to a parsable format

Reply:
I agree with you conceptually, but this is really a hard panic, something which must never happen (and wont). I generally use the following approach for tools like this:

- Any error which is consequence of user input is gracefully reported
- Internal programming errors still crash

WDYT?